### ISSUE TITLE
Error (maybe) in tappy imports

### ISSUE THREAD

**Issue by lidiabressan** [ARPA-SIMC/tappy#1]
I get this error trying to use the function tappy.tappy.TAPPY:

> Traceback (most recent call last):
>   File "utili_tides_test.py", line 30, in <module>
>     ris = const_decomposition_tappy(dati_df )
>   File "utili_tides_test.py", line 22, in const_decomposition_tappy
>     ris = tappy2.TAPPY(joined)
>   File "/autofs/nfshomes/lbressan/local3/lib64/python3.7/site-packages/tappy/tappy.py", line 1713, in TAPPY
>     x = tappy.tappy(outputts=outputts,
> AttributeError: type object 'tappy' has no attribute 'tappy'

My imports:
```
import tappy as tappy1
import tappy.tappy as tappy2
```

But there is a tappy in tappy2:

    `print(dir(tappy2))`

> ['TAPPY', 'Util', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__spec__', '__version__', '_master_speed_dict', 'baker', 'cal', 'datetime', 'debug', 'debug_p', 'deg2rad', 'fatal', 'interpolate', 'leastsq', 'letter_to_factor_map', 'modname', 'msg', 'node_factor_144', 'node_factor_149', 'node_factor_227', 'node_factor_235', 'node_factor_73', 'node_factor_74', 'node_factor_75', 'node_factor_76', 'node_factor_77', 'node_factor_78', 'node_factor_79', 'np', 'operator', 'os', 'pad', 'rad2deg', 'run_TAPPY', 'sparser', 'sys', 'tappy', 'tappy_lib', 'usage', 'zone_calculations']

Could the error come from the various internal tappy imports in the package (already corrected in the last commit)?

I installed it in a tappy directory, containing an other tappy directory and a tappy.py file:
~/git/tappy/tappy/tappy.py

Here is the code I use:

```
import tappy as tappy1
import tappy.tappy as tappy2
import numpy as np
import pandas

def const_decomposition_tappy(dati_df ):
    dates = np.array([dati_df.index.year.values,
                      dati_df.index.month.values,
                      dati_df.index.day.values, \
                      dati_df.index.hour.values,
                      dati_df.index.minute.values,
                      dati_df.index.second.values]).T
    values = dati_df.values
    joined = np.concatenate((dates, values[:,np.newaxis]), axis=1)
    #print(dir(tappy1))
    print(dir(tappy2))
    ris = tappy2.TAPPY(joined)
    return ris


if __name__=="__main__":
    tempi  = pandas.date_range('2018-11-10', periods=24*10, freq='H')
    dati_df = pandas.Series(np.sin(2*np.pi*np.arange(tempi.shape[0])/24.), index=tempi)
    ris = const_decomposition_tappy(dati_df )

```

thanks!
Lidia

**Comment by edigiacomo** [comment]
The module `tappy.tappy` is trying to use a function `tappy.tappy`, i.e. function `tappy` in module `tappy`, but that function is defined inside the `tappy.tappy` module itself. Something similar happens for `tappy.rad2deg`. I got lost after the 10th recursive tappy, so I don't know how it could work in Python2.7 :grimacing: 

The issue should be fixed in branch `python3`, please test the code and close this issue if everything's ok.

**Comment by lidiabressan** [comment]
Now my example works, but for some strange reasons all tests and examples fail.

**Comment by lidiabressan** [comment]
The reason is that they don't find tappy.py. It looks for ../tappy.py but it is in tappy/tappy.py. I'm lost too..

**Comment by lidiabressan** [comment]
Changing the path in example.sh and trying the examples brings:

Traceback (most recent call last):
  File "../tappy/tappy.py", line 49, in <module>
    from . import tappy_lib
ImportError: attempted relative import with no known parent package

**Comment by edigiacomo** [comment]
> The reason is that they don't find tappy.py. It looks for ../tappy.py but it is in tappy/tappy.py. I'm lost too..

The script `tappy.py` was moved in `tappy` module in commit 80d0b382952a2bc75bee0b0107f644ddac027e43, I think that until then the examples worked fine.

> ImportError: attempted relative import with no known parent package

Yes, because the relative imports don't work when you run a Python file as a script (e.g. `python3 myscript.py`), because the script is not considered as part of a module and then the interpreter cannot find the parent module. `tappy.py` is no more a script, then it must be run as `python3 -m tappy.tappy`, which means: "run the module using `__name__ = __main__`" (this is the reason for for https://github.com/ARPA-SIMC/tappy/blob/master/tappy/tappy.py#L1868).

I have fixed `example.sh`.

**Comment by lidiabressan** [comment]
It works for me and this probably fixes also Jacopo issue. Jacopo can you check, please, to make a pull and try calling tappy as in the example? @jalessandri

**Comment by jalessandri** [comment]
Hi all, I tried to use tappy launching this command.

`python3 -m tappy.tappy --filter doodson dummy_model_demean.txt goro_def.txt`

It works fine with python2 version while here I get these errors:

>/usr/lib64/python3.7/runpy.py:125: RuntimeWarning: 'tappy.tappy' found in sys.modules after import of package 'tappy', but prior to execution of 'tappy.tappy'; this may result in unpredictable behaviour
  warn(RuntimeWarning(msg))
Traceback (most recent call last):
  File "/usr/lib64/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib64/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/autofs/nfshomes/lbressan/local3/lib64/python3.7/site-packages/TAPPY-0.10.2-py3.7.egg/tappy/tappy.py", line 2249, in <module>
  File "/autofs/nfshomes/lbressan/local3/lib64/python3.7/site-packages/baker.py", line 888, in run
    value = self.apply(*self.parse(argv), instance=instance)
  File "/autofs/nfshomes/lbressan/local3/lib64/python3.7/site-packages/baker.py", line 866, in apply
    return cmd.fn(*newargs, **newkwargs)
  File "/autofs/nfshomes/lbressan/local3/lib64/python3.7/site-packages/TAPPY-0.10.2-py3.7.egg/tappy/tappy.py", line 2115, in analysis
  File "/autofs/nfshomes/lbressan/local3/lib64/python3.7/site-packages/TAPPY-0.10.2-py3.7.egg/tappy/tappy.py", line 1260, in filters
  File "/autofs/nfshomes/lbressan/local3/lib64/python3.7/site-packages/TAPPY-0.10.2-py3.7.egg/tappy/tappy.py", line 972, in missing
IndexError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`) and integer or boolean arrays are valid indices

Here the time series and the definition file I used:
[dummy_model_demean.txt](https://github.com/ARPA-SIMC/tappy/files/4797263/dummy_model_demean.txt)
[goro_def.txt](https://github.com/ARPA-SIMC/tappy/files/4797257/goro_def.txt)

Thank you very much for your support.

**Comment by edigiacomo** [comment]
> IndexError: only integers, slices (:), ellipsis (...), numpy.newaxis (None) and integer or boolean arrays are valid indices

The problem is that `interval = 5135`, then `interval/2 = 2567.5` and it's not a valid index. In Python 2.7, the operator `/` is an integer division, whereas in Python3 is a float division. I have fixed it in 0404eaee448dcef92c998fbb2f9a73f189e9a221

> /usr/lib64/python3.7/runpy.py:125: RuntimeWarning: 'tappy.tappy' found in sys.modules after import of package 'tappy', but prior to execution of 'tappy.tappy'; this may result in unpredictable behaviour

This is the first time that I see a warning like this. But I suddenly need to watch _Being John Malkovic_ :grimacing:

**Comment by jalessandri** [comment]
I think the major problems for python3 version are solved. I'm going to close the issue. thank you all. Now i need to watch _Being John Malkovic_. It seems a very interesting story.
